storage/pebble/procedure/executed.go

```diff
@@ -4,35 +4,36 @@ import (
 	"errors"
 	"fmt"
 
-	"github.com/dgraph-io/badger/v2"
+	"github.com/cockroachdb/pebble"
 
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/storage"
-	"github.com/onflow/flow-go/storage/badger/operation"
+	"github.com/onflow/flow-go/storage/pebble/operation"
 )
 
 // UpdateHighestExecutedBlockIfHigher updates the latest executed block to be the input block
 // if the input block has a greater height than the currently stored latest executed block.
 // The executed block index must have been initialized before calling this function.
 // Returns storage.ErrNotFound if the input block does not exist in storage.
-func UpdateHighestExecutedBlockIfHigher(header *flow.Header) func(txn *badger.Txn) error {
-	return func(txn *badger.Txn) error {
+func UpdateHighestExecutedBlockIfHigher(header *flow.Header) func(storage.PebbleReaderBatchWriter) error {
+	return func(rw storage.PebbleReaderBatchWriter) error {
+		r, tx := rw.ReaderWriter()
 		var blockID flow.Identifier
-		err := operation.RetrieveExecutedBlock(&blockID)(txn)
+		err := operation.RetrieveExecutedBlock(&blockID)(r)
 		if err != nil {
 			return fmt.Errorf("cannot lookup executed block: %w", err)
 		}
 
 		var highest flow.Header
-		err = operation.RetrieveHeader(blockID, &highest)(txn)
+		err = operation.RetrieveHeader(blockID, &highest)(r)
 		if err != nil {
 			return fmt.Errorf("cannot retrieve executed header: %w", err)
 		}
 
 		if header.Height <= highest.Height {
 			return nil
 		}
-		err = operation.UpdateExecutedBlock(header.ID())(txn)
+		err = operation.InsertExecutedBlock(header.ID())(tx)
```

Comment:
## ⚠️ 

this is always vulnerable to stale writes in case of concurrent access, which would corrupt the state. See [my related comment](https://github.com/onflow/flow-go/pull/6197#discussion_r1715978123) for details.

I think we need a refactoring here, where we move the logic to inserting blocks and indexing children to one specific component that has the sole authority to maintain these indices in a consistent manner.